### PR TITLE
Use defaultConfig for pallet_contracts

### PR DESCRIPTION
**Pull request by pgherveou**

### CODE REVIEW

File: substrate/frame/contracts/src/tests.rs [hunk 1]
```diff
@@ -457,28 +436,21 @@ parameter_types! {
 	pub static UnstableInterface: bool = true;
 }
 
+#[derive_impl(crate::config_preludes::TestDefaultConfig as crate::DefaultConfig)]
 impl Config for Test {
```
Comment (pgherveou) [inline substrate/frame/contracts/src/tests.rs:440]: Some of the default trait are overridden here as we needs to use non const values to test different scenarios

File: substrate/frame/contracts/src/lib.rs [hunk 2]
```diff
@@ -279,21 +288,25 @@ pub mod pallet {
...
 		/// construct a default cost schedule.
 		type WeightInfo: WeightInfo;
 
 		/// Type that allows the runtime authors to add new host functions for a contract to call.
+		#[pallet::no_default]
 		type ChainExtension: chain_extension::ChainExtension<Self> + Default;
 
 		/// Cost schedule and limits.
 		#[pallet::constant]
+		#[pallet::no_default]
```
Comment (pgherveou) [inline substrate/frame/contracts/src/lib.rs:309]: some of these traits that take Self can't have default I suppose...

File: substrate/frame/contracts/src/lib.rs [hunk 3]
```diff
@@ -312,10 +326,12 @@ pub mod pallet {
 		///
 		/// Changing this value for an existing chain might need a storage migration.
 		#[pallet::constant]
+		#[pallet::no_default]
 		type DepositPerByte: Get<BalanceOf<Self>>;
```
Comment (ggwpez) [inline substrate/frame/contracts/src/lib.rs:330]: Is a default really not possible here @gupnik ? Thats a bit sad...

Comment (gupnik) [reply substrate/frame/contracts/src/lib.rs:330]: You an use `pallet::no_default_bounds` here @pgherveou @ggwpez

Comment (pgherveou) [reply substrate/frame/contracts/src/lib.rs:330]: got it and then in TestDefaultConfig I have to hard code the Balance type defined here?
https://github.com/paritytech/polkadot-sdk//blob/0bf6e1ab48f2a3918a95e6001da618a2ef0ed750/substrate/frame/balances/src/lib.rs#L231
So in a way they are all tightly coupled, which I guess is ok for Test config.

How would that work if we wanted to use that for something else than well defined TestConfig?

Comment (gupnik) [reply substrate/frame/contracts/src/lib.rs:330]: Once you set `no_default_bounds`, internally it just removes all bounds and creates `type DepositPerByte` in the trait DefaultConfig. 

So, you can pretty much set it to anything as long as it makes sense within your TestDefaultConfig.

Comment (pgherveou) [reply substrate/frame/contracts/src/lib.rs:330]: is it worthwhile adding dummy value for these fields though if they are going to be overridden anyway?

**Review by sam0x17 — APPROVED** [review]

File: substrate/frame/contracts/src/lib.rs (hunk 3 above)
Comment (gupnik) [reply substrate/frame/contracts/src/lib.rs:330]: If you want to ensure that users override this value, its absolutely fine to set `no_default`. Is there no sensible default for `DepositPerByte`?

**Comment by gupnik** [comment]
@pgherveou Are we good to proceed here? Any blockers?

**Comment by pgherveou** [comment]
> @pgherveou Are we good to proceed here? Any blockers?

Kind of forgot about it, will bring it up to date :)

**Comment by pgherveou** [comment]
just merge conflicts, will re-go through the different types to pick sensible defaults

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5887491

**Comment by pgherveou** [comment]
cc @gupnik  how does that look now?

File: substrate/frame/contracts/mock-network/src/parachain/contracts_config.rs [hunk 4]
```diff
@@ -15,87 +15,19 @@
...
-	weights::Weight,
-};
-use frame_system::{pallet_prelude::BlockNumberFor, EnsureSigned};
-use pallet_xcm::BalanceOf;
-use sp_runtime::{traits::Convert, Perbill};
-
-pub const fn deposit(items: u32, bytes: u32) -> Balance {
-	items as Balance * 1 * CENTS + (bytes as Balance) * 1 * CENTS
-}
+use crate::parachain::RuntimeHoldReason;
```
Comment (pgherveou) [inline substrate/frame/contracts/mock-network/src/parachain/contracts_config.rs:18]: This need to be in scope for the derive_impl macro to work, is that expected @gupnik?

Comment (gupnik) [reply substrate/frame/contracts/mock-network/src/parachain/contracts_config.rs:18]: Yeah, correct. It assumes that these would be provided by the runtime v2 or construct_runtime macro.

File: substrate/frame/contracts/src/lib.rs [hunk 5]
```diff
@@ -250,9 +250,10 @@ pub mod pallet {
 	#[pallet::storage_version(STORAGE_VERSION)]
 	pub struct Pallet<T>(_);
 
-	#[pallet::config]
+	#[pallet::config(with_default)]
 	pub trait Config: frame_system::Config {
 		/// The time implementation used to supply timestamps to contracts through `seal_now`.
+		#[pallet::no_default]
```
Comment (gupnik) [inline substrate/frame/contracts/src/lib.rs:256]: Not possible to provide a default here?

File: substrate/frame/contracts/src/lib.rs [hunk 6]
```diff
@@ -301,21 +310,25 @@ pub mod pallet {
...
 		/// construct a default cost schedule.
 		type WeightInfo: WeightInfo;
 
 		/// Type that allows the runtime authors to add new host functions for a contract to call.
+		#[pallet::no_default_bounds]
 		type ChainExtension: chain_extension::ChainExtension<Self> + Default;
 
 		/// Cost schedule and limits.
 		#[pallet::constant]
+		#[pallet::no_default]
```
Comment (gupnik) [inline substrate/frame/contracts/src/lib.rs:331]: Tried with `no_default_bounds`?

File: substrate/frame/contracts/src/lib.rs [hunk 7]
```diff
@@ -326,6 +339,7 @@ pub mod pallet {
 		///
 		/// This setting along with [`MaxCodeLen`](#associatedtype.MaxCodeLen) directly affects
 		/// memory usage of your runtime.
+		#[pallet::no_default]
```
Comment (gupnik) [inline substrate/frame/contracts/src/lib.rs:342]: Same?

File: substrate/frame/contracts/src/tests.rs [hunk 8]
```diff
@@ -334,34 +334,24 @@ parameter_types! {
...
+	type Block = Block;
 	type AccountId = AccountId32;
 	type Lookup = IdentityLookup<Self::AccountId>;
-	type Block = Block;
 	type AccountData = pallet_balances::AccountData<u64>;
 }
+
 impl pallet_insecure_randomness_collective_flip::Config for Test {}
+
+#[derive_impl(pallet_balances::config_preludes::TestDefaultConfig)]
```
Comment (gupnik) [inline substrate/frame/contracts/src/tests.rs:345]: 👍

**Review by gupnik — APPROVED** [review]

File: substrate/frame/contracts/src/lib.rs (hunk 6 above)
Comment (pgherveou) [reply substrate/frame/contracts/src/lib.rs:331]: Nop the trait is generic over T
https://github.com/paritytech/polkadot-sdk/blob/pg/contract_default_trait/substrate/frame/contracts/src/schedule.rs?plain=1#L57-L66

File: substrate/frame/contracts/src/lib.rs (hunk 7 above)
Comment (pgherveou) [reply substrate/frame/contracts/src/lib.rs:342]: same, the trait is generic over T

File: substrate/frame/contracts/src/lib.rs (hunk 5 above)
Comment (pgherveou) [reply substrate/frame/contracts/src/lib.rs]: Yeah you would most likely want to override it, if you do anything that use that type in your tests, but I can provide an implementation that is unimplented!

**Pull request merged**